Enable adding of Service Endpoints to subnets

mtougeron opened the pull request:
**What type of PR is this?**

/kind feature

**What this PR does / why we need it**:

This implements the ability to add Service Endpoints to the subnets created/managed by CAPZ.


**Which issue(s) this PR fixes**:

Fixes https://github.com/kubernetes-sigs/cluster-api-provider-azure/issues/2621 

**TODOs**:

- [x] squashed commits
- [ ] includes documentation
- [x] adds unit tests

**Release note**:
```release-note
Enables adding of Service Endpoints to subnets created/managed by CAPZ
```

mtougeron commented:
Current outstanding TODOs:
- [ ] Write new docs on how to use this features along with limitations
- [ ] Refactor the subnet reconcile function so it can update existing subnets with service endpoint changes
- [ ] fix the api version conversion logic

mtougeron commented on exp/api/v1alpha3/azuremanagedcontrolplane_conversion.go:
```diff
@@ -81,3 +82,8 @@ func (dst *AzureManagedControlPlaneList) ConvertFrom(srcRaw conversion.Hub) erro
 }
+
+// Convert_v1beta1_ManagedControlPlaneSubnet_To_v1alpha3_ManagedControlPlaneSubnet is an autogenerated conversion function.
+func Convert_v1beta1_ManagedControlPlaneSubnet_To_v1alpha3_ManagedControlPlaneSubnet(in *infrav1exp.ManagedControlPlaneSubnet, out *ManagedControlPlaneSubnet, s apiconversion.Scope) error {
```
I know this is wrong, but I'm not sure how to do it right. Without it, it gives

```
E0907 12:05:50.092587    9028 conversion.go:756]   the following fields need manual conversion:
E0907 12:05:50.092592    9028 conversion.go:758]       - ServiceEndpoints
```

but with it, it's a duplicate declaration when trying to build the app.

jantzenallphin commented on api/v1beta1/azurecluster_validation_test.go:
```diff
@@ -1378,3 +1378,130 @@ func createValidAPIServerInternalLB() LoadBalancerSpec {
+	}
+
+	testCase := test{
+		name: "service endpoint missing service name",
```
This test name should be something like `name: "service endpoint missing locations",`

CecileRobertMichon commented:
/ok-to-test

mtougeron commented:
Per the [slack conversation](https://kubernetes.slack.com/archives/CEX9HENG7/p1662487418501269), the sticking point I'm currently running into is this `make generate` error. Otherwise things seem to be working.

```
/Users/tougeron/src/go/src/github.com/mtougeron/cluster-api-provider-azure/hack/tools/bin/conversion-gen-v0.23.1 \
		--input-dirs=./exp/api/v1alpha3 \
		--output-file-base=zz_generated.conversion \
		--go-header-file=./hack/boilerplate/boilerplate.generatego.txt --output-base=/Users/tougeron/src/go/src/github.com/mtougeron/cluster-api-provider-azure
E0906 08:11:50.114503   66080 conversion.go:755] Warning: could not find nor generate a final Conversion function for sigs.k8s.io/cluster-api-provider-azure/exp/api/v1beta1.ManagedControlPlaneSubnet -> github.com/mtougeron/cluster-api-provider-azure/exp/api/v1alpha3.ManagedControlPlaneSubnet
E0906 08:11:50.114678   66080 conversion.go:756]   the following fields need manual conversion:
E0906 08:11:50.114686   66080 conversion.go:758]       - ServiceEndpoints
```


Also need to figure out a refactoring for:

```
azure/services/subnets/spec.go:60:1: cyclomatic complexity 36 of func `(*SubnetSpec).Parameters` is high (> 30) (gocyclo)
func (s *SubnetSpec) Parameters(existing interface{}) (parameters interface{}, err error) {
```

mtougeron commented on azure/services/subnets/spec.go:
```diff
@@ -54,15 +56,90 @@ func (s *SubnetSpec) OwnerResourceName() string {
 }
 
+// verifyRouteTableImmutable validates that an immutable change isn't made to the subnet's route table.
+func (s *SubnetSpec) verifyRouteTableImmutable(existingSubnet network.Subnet) error {
```
I'm not happy with these function names. If you have a better suggestion I'm happy to implement them. It's basically meant to validate that the changes are not made to those fields since they are immutable. I originally had them in the `Parameters` function but it was failing the cyclomatic complexity checks.

CecileRobertMichon replied:
why is this validation happening here and not in the webhook?

mtougeron replied:
Sounds like you just answered https://github.com/kubernetes-sigs/cluster-api-provider-azure/pull/2635#issuecomment-1261222657 for me. :)

jackfrancis replied:
Yes, thanks Cecile, sorry I didn't catch this earlier. We want these validations to fire before the resource updates ever enter into the reconciliation.

mtougeron replied:
This has been updated

mtougeron commented:
The more detailed apidiff is

```
sigs.k8s.io/cluster-api-provider-azure/api/v1beta1
  Compatible changes:
  - ServiceEndpointSpec: added
  - ServiceEndpoints: added
  - SubnetSpec.ServiceEndpoints: added

sigs.k8s.io/cluster-api-provider-azure/azure/services/subnets
  Compatible changes:
  - SubnetSpec.ServiceEndpoints: added

sigs.k8s.io/cluster-api-provider-azure/exp/api/v1beta1
  Incompatible changes:
  - ManagedControlPlaneSubnet: old is comparable, new is not
  - ManagedControlPlaneVirtualNetwork: old is comparable, new is not
  Compatible changes:
  - ManagedControlPlaneSubnet.ServiceEndpoints: added
```

but I'm not sure why it works fine for the AzureCluster but not ManagedControlPlane.

mtougeron commented:
/retest

jackfrancis commented on azure/scope/cluster.go:
```diff
@@ -379,12 +379,19 @@ func (s *ClusterScope) SubnetSpecs() []azure.ResourceSpecGetter {
 			NatGatewayName:    subnet.NatGateway.Name,
 		}
+		if s.IsVnetManaged() {
+			serviceEndpointList := make([]infrav1.ServiceEndpointSpec, 0, len(subnet.ServiceEndpoints))
```
Let's do this `serviceEndpointList` construction in a separate func as it appears to be identical in L405-413 below.

mtougeron replied:
done

mtougeron commented on api/v1beta1/azurecluster_validation.go:
```diff
@@ -51,6 +51,10 @@ const (
 	minRulePriority = 100
 	maxRulePriority = 4096
+	// Must start with 'Microsoft.', then an alpha character, then can include alnum.
+	serviceEndpointServiceRegex = `^Microsoft.[a-zA-Z]{1,42}[a-zA-Z0-9]{0,42}$`
```
darn regex! Fixed. I'll push after waiting for any other feedback.

mtougeron commented on api/v1beta1/azurecluster_validation.go:
```diff
@@ -548,3 +556,55 @@ func validateClassSpecForControlPlaneOutboundLB(lb *LoadBalancerClassSpec, apise
+}
+
+func validateServiceEndpointServiceName(serviceName string, fldPath *field.Path) *field.Error {
+	regexpService := regexp.MustCompile(serviceEndpointServiceRegex)
```
pre-compiled the regex into a var at the package level.

mtougeron commented:
@nojnhuh / @jackfrancis I'm not sure what the apidiff job is telling me is wrong. I know it's tell me _something_ is wrong but I don't know how to track down what it is since the ManagedControlPlane changes are the same as the AzureCluster changes (as far as I can tell).

nojnhuh commented:
This is about 50/50 spitball/educated guess, but I think apidiff is complaining that these changes remove the ability to meaningfully compare those types with `==` since `ManagedControlPlaneSubnet` only contained two strings before but now includes a slice. I imagine it's not complaining about AzureCluster because those types were already incomparable before these changes.

jackfrancis commented on api/v1beta1/types.go:
```diff
@@ -589,9 +592,20 @@ type SubnetSpec struct {
 
+	// ServiceEndpoints is a slice of Virtual Network service endpoints to enable for the subnets.
+	// +optional
+	ServiceEndpoints `json:"serviceEndpoints,omitempty"`
```
Is there an argument that this is more clear?:

```
// ServiceEndpoints is a slice of Virtual Network service endpoints to enable for the subnets.
// +optional
ServiceEndpoints ServiceEndpoints `json:"serviceEndpoints,omitempty"`
```

@nojnhuh @CecileRobertMichon  any thoughts here? A cursory look suggests that other slice types in this file (`Subnets`, `VnetPeerings`) are not embedded like the above, but have an explicit property name when used in other type definitions.

AFAIK in golang the two options are exactly equivalent, but for my combining the golang struct embedding pattern w/ slice types is a bit harder to grok.

nojnhuh replied:
I tend to agree that if we don't need the hoisting-like attributes of embedded properties, then giving an explicit name might be a bit more clear. It might not be as big of a deal in this particular since Go doesn't let you [index into a type that embeds a slice](https://go.dev/play/p/b7HYbUtuTkm) which would likely be surprising to read, but I can see value in naming the field for consistency's sake.

mtougeron replied:
This is an oversight on my part. I meant to do `ServiceEndpoints ServiceEndpoints `json:"serviceEndpoints,omitempty"`` like I did for the managed control plane.

jackfrancis commented on azure/scope/cluster.go:
```diff
@@ -379,12 +379,15 @@ func (s *ClusterScope) SubnetSpecs() []azure.ResourceSpecGetter {
 			Role:              subnet.Role,
 			NatGatewayName:    subnet.NatGateway.Name,
 		}
+		if s.IsVnetManaged() {
```
Do we want to check if the length of `subnet.ServiceEndpoints` is greater than zero before entering into this func and performing `make([]infrav1.ServiceEndpointSpec, 0, len(subnetServiceEndpoints))`? Doing it this way essentially assigns the zero value to `subnetSpec.ServiceEndpoints` when there are no `ServiceEndpoints` defined, but at an extra cost.

mtougeron replied:
It needs to be able to be set to an empty slice in the case of someone switching from having service endpoints defined to not having any service endpoints defined. If I add the conditional it won't remove the service endpoints in this scenario.

jackfrancis replied:
Thanks for thinking of all cluster configuration mutation scenarios!

jackfrancis commented on azure/scope/cluster.go:
```diff
@@ -379,12 +379,15 @@ func (s *ClusterScope) SubnetSpecs() []azure.ResourceSpecGetter {
 	if s.IsAzureBastionEnabled() {
 		azureBastionSubnet := s.AzureCluster.Spec.BastionSpec.AzureBastion.Subnet
-		subnetSpecs = append(subnetSpecs, &subnets.SubnetSpec{
+		subnetSpec := &subnets.SubnetSpec{
```
slight preference for renaming the temporary `subnetSpec` var here to `azureBastionSubnetSpec` to aid the developer's eye in the future when maintaining this as distinct from the above, similar stuff in the `s.AzureCluster.Spec.NetworkSpec.Subnets` enumeration (L369)

mtougeron replied:
updated

jackfrancis commented on azure/scope/cluster.go:
```diff
@@ -395,7 +398,11 @@ func (s *ClusterScope) SubnetSpecs() []azure.ResourceSpecGetter {
 			Role:              azureBastionSubnet.Role,
-		})
+		}
+		if s.IsVnetManaged() {
```
ditto above about maybe checking if there actually are any `azureBastionSubnet.ServiceEndpoints` before potentially assigning a strongly-typed nil value to `subnetSpec.ServiceEndpoints` unnecessarily

jackfrancis commented on azure/scope/managedcontrolplane.go:
```diff
@@ -258,18 +258,29 @@ func (s *ManagedControlPlaneScope) NodeNatGateway() infrav1.NatGateway {
+
+	if s.IsVnetManaged() {
+		var serviceEndpointList []infrav1.ServiceEndpointSpec
+		for _, se := range s.NodeSubnet().ServiceEndpoints {
```
ditto ditto above

jackfrancis reviewed (approved):
/lgtm

@mtougeron we don't have a lot of E2E coverage around BYO VNET for AzureManagedCluster scenarios, are you confident in your own manual tests that this is feature complete and ready to use? code lgtm

Thank you so much!

mtougeron commented:
> @mtougeron we don't have a lot of E2E coverage around BYO VNET for AzureManagedCluster scenarios, are you confident in your own manual tests that this is feature complete and ready to use? code lgtm

I'll run through another round of manual testing today to be doubly sure. Is there any sort of logging I can provide that would make you feel confident in the testing as well?

mtougeron commented:
@jackfrancis I tested this again for BYO VNET and it doesn't make any changes (as expected/desired).

However, as I look closer, there isn't any validation done on the `AzureManagedControlPlane.spec.virtualNetwork` (and sub fields like `subnet`) in the `exp/api/v1beta1/azuremanagedcontrolplane_webhook.go` code. Would you like me to add that new logic now or should I wait for it to be refactored to be more like the `api/v1beta1/azurecluster_validation.go` code?

As this relates to this PR, it just ignores the new serviceEndpoints field with BYO VNET and doesn't report anything to the user. While this is not the expected experience, it is similar to what happens when a user sets other fields that are not used with BYO VNET.

CecileRobertMichon commented on azure/scope/cluster.go:
```diff
@@ -1004,3 +1011,11 @@ func (s *ClusterScope) TagsSpecs() []azure.TagsSpec {
 	}
 }
+
+func buildServiceEndpointList(subnetServiceEndpoints infrav1.ServiceEndpoints) []infrav1.ServiceEndpointSpec {
```
I'm not sure why we need this extra func here, isn't `infrav1.ServiceEndpoints` equivalent to `[]infrav1.ServiceEndpointSpec`?

mtougeron replied:
Heh, good point. This originally started because of a field being a pointer (see the subnet/spec.go) but I changed that an never came back and refactored this code. I'll remove this.

CecileRobertMichon commented on azure/scope/cluster.go:
```diff
@@ -395,7 +398,11 @@ func (s *ClusterScope) SubnetSpecs() []azure.ResourceSpecGetter {
 			Role:              azureBastionSubnet.Role,
-		})
+		}
+		if s.IsVnetManaged() {
```
Why the differentiation between BYO vnet and managed vnet here? The entire reconciliation should be skipped if the vnet is not managed, and this differs from what we're doing for the other fields

mtougeron replied:
If the entire reconciliation is skipped if it isn't managed then this conditional can be removed. I'll update

mtougeron commented:
/retest

mtougeron commented:
/retest

mtougeron commented:
Apparently I created a regression with my last round of fixes since the `pull-cluster-api-provider-azure-e2e` test keeps failing. I'll figure out what's different in my manual testing vs the automated test suites and get that fixed.

mtougeron commented:
/test pull-cluster-api-provider-azure-e2e

mtougeron commented:
Ah, finally tracked down the failing log message.

```
E0930 16:32:16.616238       1 controller.go:326]  "msg"="Reconciler error" "error"="admission webhook \"validation.azurecluster.infrastructure.cluster.x-k8s.io\" denied the request: AzureCluster.infrastructure.cluster.x-k8s.io \"capz-e2e-i7wx2l-public-custom-vnet\" is invalid: [spec.networkSpec.subnets[0].CIDRBlocks: Invalid value: []string{\"10.0.0.0/24\"}: field is immutable, spec.networkSpec.subnets[1].CIDRBlocks: Invalid value: []string{\"10.0.1.0/24\"}: field is immutable]" "azureCluster"={"name":"capz-e2e-i7wx2l-public-custom-vnet","namespace":"capz-e2e-i7wx2l"} "controller"="azurecluster" "controllerGroup"="infrastructure.cluster.x-k8s.io" "controllerKind"="AzureCluster" "name"="capz-e2e-i7wx2l-public-custom-vnet" "namespace"="capz-e2e-i7wx2l" "reconcileID"="2afbcc3e-fa11-444f-9bf3-019cc147134f" 
```

Looks like the spec gets updated with the cidr blocks from the custom vnet but my checks for changes to those fields are preventing that. Previously the validation only did the checks if it was a managed vnet which is why it was passing previously. I'll update and push the fix

mtougeron commented:
/test pull-cluster-api-provider-azure-e2e

mtougeron commented:
Not sure what's up with prow but I'm going to try again.

/test pull-cluster-api-provider-azure-e2e

mtougeron commented on exp/api/v1beta1/azuremanagedcontrolplane_webhook.go:
```diff
@@ -240,6 +240,24 @@ func (m *AzureManagedControlPlane) ValidateUpdate(oldRaw runtime.Object, client
 		}
 	}
 
+	// NOTE: This only works because we force the user to set the CIDRBlock for both the
```
@jackfrancis @CecileRobertMichon Please note this caveat with this logic. Personally I consider the lack of reconciliation of the Subnet cidr block a bug and think it should be done like the AzureCluster, but I think for this purpose we should cross that bridge when it happens. Thoughts?

jackfrancis replied:
Thanks for your stamina here @mtougeron, will get to a thorough review of this soon. ❤️

mtougeron commented on api/v1beta1/azurecluster_webhook.go:
```diff
@@ -128,6 +128,44 @@ func (c *AzureCluster) ValidateUpdate(oldRaw runtime.Object) error {
+			// Verify the CIDR blocks haven't changed for an owned Vnet.
+			// A non-owned Vnet's CIDR block can change based on what's
+			// defined in the spec vs what's been loaded from Azure directly.
+			if old.Spec.NetworkSpec.Vnet.Tags.HasOwned(old.Name) && !reflect.DeepEqual(subnet.CIDRBlocks, oldSubnet.CIDRBlocks) {
```
This technically allows the cidr block to be modified in the brief moments before the Vnet is created (because the tags haven't been set yet) but once the Vnet has been created it becomes immutable. That seems acceptable to me since it doesn't create any changes in the subnet since it doesn't exist yet. Thoughts?

mtougeron reviewed (commented):
@jackfrancis @CecileRobertMichon @nojnhuh This is ready for review again. I got the validation of the immutability working for the subnets. Thanks!

mtougeron commented:
Added the SSA flags for the new field

mtougeron commented:
Prow flake

/test pull-cluster-api-provider-azure-test

CecileRobertMichon commented on api/v1beta1/azurecluster_webhook.go:
```diff
@@ -128,6 +128,44 @@ func (c *AzureCluster) ValidateUpdate(oldRaw runtime.Object) error {
 		)
 	}
 
+	oldSubnetMap := make(map[string]SubnetSpec, len(old.Spec.NetworkSpec.Subnets))
```
could we extract this subnet immutability validation to its own function?

mtougeron replied:
Sure, no problem.

CecileRobertMichon commented on api/v1beta1/azurecluster_webhook.go:
```diff
@@ -128,6 +128,44 @@ func (c *AzureCluster) ValidateUpdate(oldRaw runtime.Object) error {
+			// Verify the CIDR blocks haven't changed for an owned Vnet.
+			// A non-owned Vnet's CIDR block can change based on what's
+			// defined in the spec vs what's been loaded from Azure directly.
+			if old.Spec.NetworkSpec.Vnet.Tags.HasOwned(old.Name) && !reflect.DeepEqual(subnet.CIDRBlocks, oldSubnet.CIDRBlocks) {
```
how does this work when the tags haven't been set yet (they get set by the controller when we create/get the vnet the first time) we're creating a new cluster but the CIDRBlock gets defaulted in the defaulting webhook? Should this allow going from empty string to not-empty-string?

CecileRobertMichon replied:
ha just read https://github.com/kubernetes-sigs/cluster-api-provider-azure/pull/2635#discussion_r985268538 which basically answers my question

Could you please clarify this in a code comment? That's good info. And to answer your question, I agree that's acceptable too, as long as we're not blocking the webhook from defaulting.

mtougeron replied:
Sure, I'll make the comment more verbose. Thanks!

mtougeron replied:
Added

CecileRobertMichon commented on api/v1beta1/types.go:
```diff
@@ -617,9 +622,20 @@ type SubnetSpec struct {
 
+	// ServiceEndpoints is a slice of Virtual Network service endpoints to enable for the subnets.
+	// +optional
+	ServiceEndpoints ServiceEndpoints `json:"serviceEndpoints,omitempty"`
```
Is this something that would make sense to reuse across multiple clusters when creating a "fleet" of clusters via ClusterClass? If so it should go into `SubnetClassSpec`

mtougeron replied:
Honestly I'm not sure but I don't think it would hurt. I'll move it.

mtougeron replied:
Updated.

CecileRobertMichon commented on docs/book/src/topics/custom-vnet.md:
````diff
@@ -176,6 +176,43 @@ spec:
   resourceGroup: cluster-example
 ```
 
+### Virtual Network service endpoints
````
thank you for documenting this!

CecileRobertMichon commented:
@mtougeron thanks for your patience with reviews here. This is a big and complex change, but FWIW I think we're really close. Just have a few more questions from my side.

mtougeron commented:
> @mtougeron thanks for your patience with reviews here. This is a big and complex change, but FWIW I think we're really close. Just have a few more questions from my side.

No problem at all. I've learned a heck of a lot in this one. It's been a ton of fun. :)

CecileRobertMichon reviewed (commented):
/lgtm

jackfrancis commented:
/approve
